### DISCUSSION TITLE
Support spinning of R scripts

### DISCUSSION THREAD

**Discussion by holgerbrandl** [quarto-dev/quarto-cli#4969] [Q&A] [answered]
A great way to render reports from R is via knitr::spin. See https://bookdown.org/yihui/rmarkdown-cookbook/spin.html for an overview.

I had previously supported a small utility to simplify spinning of R-scripts from a terminal via https://github.com/holgerbrandl/rendr directly written in R, but it always lacked beauty.

It would be great if quarto could support spinning via
```bash
quarto render my_script.R 
```
Currently, this fails with 
```
λ quarto render exp_fit.R
ERROR: Unable to render exp_fit.R

Error: Unable to render exp_fit.R
    at fileExecutionEngineAndTarget (file:///C:/Program%20Files/Quarto/bin/quarto.js:57992:15)
    at renderFiles (file:///C:/Program%20Files/Quarto/bin/quarto.js:86443:52)
```

What's your take on that?

**Comment by mcanouil** [comment]
Simply use `quarto run script.R`.
<img width="1624" alt="image" src="https://user-images.githubusercontent.com/8896044/227740856-11201119-4faf-45be-8b12-89a52297bd53.png">

**Comment by holgerbrandl** [comment]
Thanks @mcanouil . I've tried to do so, but for me this creates just a `Rplots.pdf` only, so it seems. that it's not spinning the script but rather doing something like `R -f <script>` internally.

**Reply by mcanouil** [reply]
Oh, so you've got an R script that contains roxygen2 style text.
This is highly specific to R.

I don't believe Quarto can (or will) do that because Quarto is running R, not the other way around, and for spinning to work, knitr/roxygen2 must be used first to parse the script and detect what's code and what's typesetting, then send both part to engine.

**Reply by holgerbrandl** [reply]
It's sufficient to run R from Quarto, because all it takes to spin a document it so run `knitr::spin(inputfile, text=outputfile knit=FALSE)` to create an Rmarkdown Rmd file `outputfile`, which could subsequently be run through the existing Rmd workflow of quarto. See also https://www.rdocumentation.org/packages/knitr/versions/1.42/topics/spin

**Reply by mcanouil** [reply]
The issue is that you want to include format and other option in your R script (the YAML header), but it's too late in the process if you run this command from Quarto.

But you can easily do the two steps yourself and maybe the two steps could somehow become internal to Quarto CLI when input to `quarto render` is an R script:

```sh
Rscript -e 'knitr::spin(inputfile, text = "demo.qmd", knit = FALSE)' && quarto render demo.qmd
```

**Reply by kylebutts** [reply]
@mcanouil This doesn't work. text is a way of providing `.R` file source code (so it classes with inputfile). I think what's being desired is an alternate `format` option to produce a .qmd file instead of a `.Rmd` file which would require change to `knitr::spin`?

**Reply by mcanouil** [reply]
The initial spin command is not mine, but the idea remains the same.

```bash
Rscript -e 'file.copy(knitr::spin("index.R", knit = FALSE), "demo.qmd", overwrite = TRUE)' && quarto render demo.qmd'
```

**Reply by holgerbrandl** [reply]
Automating away the Rscript call would be exactly what users of knitr::spin would appreciate doing with quarto. I agree with @kylebutts, that this would indeed require a format parameter to indicate the type of processing (raw R with just a plots pdf as output) or the spun  copy. To me, spinning would even seem the better default to turn any R script into a report (html, docx, or similar).

**Reply by baptiste** [reply]
One could perhaps argue that `quarto convert` could do the spinning when provided with a R file (and perhaps do the two-way spin<->purl thing, similar to the ipynb<->qmd process for notebooks, although I don't know if there's a reverse process for spin: purl only grabs code chunks).

**Reply by mcanouil** [reply]
Note that a Jupyter Notebook is not a Python script.
For instance, there is nothing that can convert a Python script into a Quarto document which would be the correct parallel with `knitr::spin` for R scripts.

Quarto's main focus is on writing content, thus R/Python scripts are not the best input for that in my opinion.
I never saw in the wild a manuscript written from an R script via the `knitr::spin` function.
It seems to be really a niche that can easily be address for simple markup via the updated command I shared.

**Reply by cderv** [reply]
@holgerbrandl wouldn't using your tool with `quarto run` provide some help ? 

````r
quarto run rend.R MyScript.R
```` 

Maybe with some tweaks. 

> I think what's being desired is an alternate format option to produce a .qmd file instead of a .Rmd file which would require change to knitr::spin?

@kylebutts I opened an issue with this suggestion: 

* https://github.com/yihui/knitr/issues/2284

**Reply by holgerbrandl** [reply]
Thanks for pushing the idea forward.

Regarding the use of rend.R. Being the author myself, nobody is offended when stating that rend.R is just a poor workaround to enable _some_ bash method to render any R-script into a html (or whatever pandoc-supported) format. Having a more structured way would be certainly better compared to what 

Adding qmd output support to knitr::spin seems a great idea to me. Where can I find a listing of format differences between .qmd and .Rmd? From what I read they seem quite similar at first glance. In my experience, users of spin rely quite a lot on chunk options for different section (e.g. to adjust image size), so it would be important to preserve this feature.

**Reply by cderv** [reply, marked as answer]
Now there is more broader issue to follow for this feature in Quarto 

* https://github.com/quarto-dev/quarto-cli/issues/6660

We'll see what can be done.

**Comment by cderv** [comment]
Related to @kylebutts suggestion in https://github.com/quarto-dev/quarto-cli/discussions/4969#discussioncomment-6792711, now **knitr** (from 1.44) supporting `spin(format = "qmd", knit = FALSE)` to transform a R script to a .qmd document. 

See 
* https://github.com/yihui/knitr/issues/2284

**Reply by holgerbrandl** [reply]
Thanks for implementing it in v1.4. Just gave it a first try, and it works just great.
